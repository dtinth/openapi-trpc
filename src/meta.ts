import { OpenAPIV3 } from 'openapi-types'

function defineKeys<T extends keyof OpenAPIV3.OperationObject>(keys: T[]): T[] {
  return keys
}

export const allowedOperationKeys = defineKeys([
  'deprecated',
  'description',
  'externalDocs',
  'summary',
  'tags',
])

/**
 * @public
 */
export interface OperationMeta
  extends Pick<
    OpenAPIV3.OperationObject,
    typeof allowedOperationKeys[number]
  > {}
