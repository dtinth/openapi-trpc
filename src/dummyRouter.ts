import { initTRPC } from '@trpc/server'
import { z } from 'zod'
import { OperationMeta } from './meta'

export const createDummyRouter = (
  t = initTRPC.meta<OperationMeta>().create(),
) => {
  return t.router({
    hello: t.router({
      world: t.procedure
        .meta({ description: 'ok' })
        .input(z.object({ name: z.string() }))
        .output(z.string())
        .query(() => 'hello world'),
    }),
  })
}

export type DummyRouter = ReturnType<typeof createDummyRouter>
export type DummyProcedure = ReturnType<
  typeof createDummyRouter
>['_def']['procedures']['hello']['_def']['procedures']['world']['_def']
