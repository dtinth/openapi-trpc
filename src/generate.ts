import { zodToJsonSchema } from 'zod-to-json-schema'
import { DummyProcedure, DummyRouter } from './dummyRouter'
import { z, AnyZodObject, ZodType } from 'zod'
import { OpenAPIV3 } from 'openapi-types'
import { OperationMeta, allowedOperationKeys } from './meta'
import { RootConfig, Router, RouterDef } from '@trpc/server'

/**
 * @public
 */
export function generateOpenAPIDocumentFromTRPCRouter<R extends Router<any>>(
  inRouter: R,
  options: GenerateOpenAPIDocumentOptions<MetaOf<R>> = {},
) {
  const router: DummyRouter = inRouter as unknown as DummyRouter
  const procs = router._def.procedures
  const paths: OpenAPIV3.PathsObject = {}
  const processOperation = (
    op: OpenAPIV3.OperationObject,
    meta: MetaOf<R>,
  ): OpenAPIV3.OperationObject => {
    return options.processOperation?.(op, meta) || op
  }
  for (const [procName, proc] of Object.entries(procs)) {
    const procDef = proc._def as unknown as DummyProcedure
    const input = procDef.inputs
      .slice(1)
      .reduce<AnyZodObject>(
        (acc, cur) => asZodObject(acc).merge(asZodObject(cur)),
        asZodObject(procDef.inputs[0] || z.object({})),
      )
    const output = procDef.output
    const inputSchema = toJsonSchema(input)
    const outputSchema = output ? toJsonSchema(asZodType(output)) : undefined
    const key = [
      '',
      ...(options.pathPrefix || '/').split('/').filter(Boolean),
      procName,
    ].join('/')
    const responses = {
      200: {
        description: (output && asZodType(output).description) || '',
        ...(outputSchema
          ? {
              content: {
                'application/json': {
                  schema: outputSchema as any,
                },
              },
            }
          : {}),
      },
    }
    const operationInfo: Partial<OpenAPIV3.OperationObject> = {
      tags: procName.split('.').slice(0, -1).slice(0, 1),
    }
    for (const key of allowedOperationKeys) {
      const value = procDef.meta?.[key]
      if (value) {
        operationInfo[key] = value as any
      }
    }
    if (procDef.query) {
      paths[key] = {
        get: processOperation(
          {
            ...operationInfo,
            operationId: procName,
            responses,
            parameters: [
              {
                in: 'query',
                name: 'input',
                content: {
                  'application/json': {
                    schema: inputSchema as any,
                  },
                },
              },
            ],
          },
          procDef.meta as any,
        ),
      }
    } else {
      paths[key] = {
        post: processOperation(
          {
            ...operationInfo,
            operationId: procName,
            responses,
            requestBody: {
              content: {
                'application/json': {
                  schema: inputSchema as any,
                },
              },
            },
          },
          procDef.meta as any,
        ),
      }
    }
  }
  const api: OpenAPIV3.Document = {
    openapi: '3.0.0',
    info: {
      title: 'tRPC HTTP-RPC',
      version: '',
    },
    paths,
  }
  return api
}

function getZodTypeName(input: unknown) {
  return (input as { _def?: { typeName?: string } } | undefined)?._def?.typeName
}

function asZodObject(input: unknown) {
  if (getZodTypeName(input) !== 'ZodObject') {
    throw new Error('Expected a ZodObject, received: ' + String(input))
  }
  return input as AnyZodObject
}

function asZodType(input: unknown) {
  if (!getZodTypeName(input)) {
    throw new Error('Expected a Zod schema, received: ' + String(input))
  }
  return input as ZodType
}

/**
 * @public
 */
export interface GenerateOpenAPIDocumentOptions<M extends OperationMeta> {
  pathPrefix?: string
  processOperation?: (
    operation: OpenAPIV3.OperationObject,
    meta: M | undefined,
  ) => OpenAPIV3.OperationObject | void
}

function toJsonSchema(input: ZodType) {
  const { $schema, ...output } = zodToJsonSchema(input)
  return output
}

type MetaOf<R extends Router<any>> = R extends Router<RouterDef<infer D, any>>
  ? D extends RootConfig<infer C>
    ? C['meta']
    : never
  : never
