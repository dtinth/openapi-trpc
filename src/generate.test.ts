import { initTRPC } from '@trpc/server'
import fs from 'fs'
import { z } from 'zod'
import { createDummyRouter } from './dummyRouter'
import { generateOpenAPIDocumentFromTRPCRouter } from './generate'
import { OperationMeta } from './meta'

it('works', () => {
  const t = initTRPC.meta<OperationMeta>().create()
  const router = t.router({
    example: t.router({
      createGreeting: t.procedure
        .meta({ summary: 'Creates a greeting' })
        .input(
          z
            .object({
              name: z.string().optional().describe('The name to greet'),
            })
            .describe('Input for createGreeting'),
        )
        .output(z.string().describe('The greeting text'))
        .mutation(() => 'ok'),
      hello: t.procedure
        .input(z.object({ text: z.string() }))
        .query(() => null),
      getAll: t.procedure
        .output(z.array(z.object({ id: z.string() })))
        .query(() => []),
    }),
    dummy: createDummyRouter(t),
  })
  const doc = generateOpenAPIDocumentFromTRPCRouter(router, {
    pathPrefix: '/trpc',
  })
  fs.mkdirSync('temp/examples', { recursive: true })
  fs.writeFileSync('temp/examples/basic.json', JSON.stringify(doc, null, 2))
  expect(doc).toMatchSnapshot()
})

it('works with array', () => {
  const t = initTRPC.meta<OperationMeta>().create()
  const router = t.router({
    exampleWithArrayAsInput: t.router({
      queryWithArrayInput: t.procedure
        .input(z.array(z.string()))
        .query(() => null),
    }),
    dummy: createDummyRouter(t),
  })
  const doc = generateOpenAPIDocumentFromTRPCRouter(router, {
    pathPrefix: '/trpc',
  })
  fs.mkdirSync('temp/examples', { recursive: true })
  fs.writeFileSync('temp/examples/array.json', JSON.stringify(doc, null, 2))
  expect(doc).toMatchSnapshot()
})

it('lets you post-process each operation, giving typed access to the meta', () => {
  interface AppMeta extends OperationMeta {
    requiresAuth?: boolean
  }
  const t = initTRPC.meta<AppMeta>().create()
  const router = t.router({
    public: t.procedure.query(() => null),
    private: t.procedure.meta({ requiresAuth: true }).query(() => null),
  })
  const doc = generateOpenAPIDocumentFromTRPCRouter(router, {
    pathPrefix: '/trpc',
    processOperation: (op, meta) => {
      if (meta?.requiresAuth) {
        op.security = [{ bearerAuth: [] }]
      }
    },
  })
  fs.mkdirSync('temp/examples', { recursive: true })
  fs.writeFileSync(
    'temp/examples/operation-processing.json',
    JSON.stringify(doc, null, 2),
  )
  expect(doc).toMatchSnapshot()
})
