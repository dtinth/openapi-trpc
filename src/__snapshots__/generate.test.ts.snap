// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lets you post-process each operation, giving typed access to the meta 1`] = `
Object {
  "info": Object {
    "title": "tRPC HTTP-RPC",
    "version": "",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/trpc/private": Object {
      "get": Object {
        "operationId": "private",
        "parameters": Array [
          Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "additionalProperties": false,
                  "properties": Object {},
                  "type": "object",
                },
              },
            },
            "in": "query",
            "name": "input",
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "",
          },
        },
        "security": Array [
          Object {
            "bearerAuth": Array [],
          },
        ],
        "tags": Array [],
      },
    },
    "/trpc/public": Object {
      "get": Object {
        "operationId": "public",
        "parameters": Array [
          Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "additionalProperties": false,
                  "properties": Object {},
                  "type": "object",
                },
              },
            },
            "in": "query",
            "name": "input",
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "",
          },
        },
        "tags": Array [],
      },
    },
  },
}
`;

exports[`works 1`] = `
Object {
  "info": Object {
    "title": "tRPC HTTP-RPC",
    "version": "",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/trpc/dummy.hello.world": Object {
      "get": Object {
        "description": "ok",
        "operationId": "dummy.hello.world",
        "parameters": Array [
          Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "additionalProperties": false,
                  "properties": Object {
                    "name": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "name",
                  ],
                  "type": "object",
                },
              },
            },
            "in": "query",
            "name": "input",
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "",
          },
        },
        "tags": Array [
          "dummy",
        ],
      },
    },
    "/trpc/example.createGreeting": Object {
      "post": Object {
        "operationId": "example.createGreeting",
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "additionalProperties": false,
                "description": "Input for createGreeting",
                "properties": Object {
                  "name": Object {
                    "description": "The name to greet",
                    "type": "string",
                  },
                },
                "type": "object",
              },
            },
          },
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "description": "The greeting text",
                  "type": "string",
                },
              },
            },
            "description": "The greeting text",
          },
        },
        "summary": "Creates a greeting",
        "tags": Array [
          "example",
        ],
      },
    },
    "/trpc/example.getAll": Object {
      "get": Object {
        "operationId": "example.getAll",
        "parameters": Array [
          Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "additionalProperties": false,
                  "properties": Object {},
                  "type": "object",
                },
              },
            },
            "in": "query",
            "name": "input",
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "items": Object {
                    "additionalProperties": false,
                    "properties": Object {
                      "id": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "id",
                    ],
                    "type": "object",
                  },
                  "type": "array",
                },
              },
            },
            "description": "",
          },
        },
        "tags": Array [
          "example",
        ],
      },
    },
    "/trpc/example.hello": Object {
      "get": Object {
        "operationId": "example.hello",
        "parameters": Array [
          Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "additionalProperties": false,
                  "properties": Object {
                    "text": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "text",
                  ],
                  "type": "object",
                },
              },
            },
            "in": "query",
            "name": "input",
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "",
          },
        },
        "tags": Array [
          "example",
        ],
      },
    },
  },
}
`;
